[BPF] Support for `DW_TAG_variant_part` in BTF generation

vadorovsky opened the pull request:
Variant part, represented by `DW_TAG_variant_part` is a structure with a discriminant and different variants, from which only one can be active and valid at the same time. The discriminant is the main difference between variant parts and unions represented by `DW_TAG_union` type.

Variant parts are used by Rust enums, which look like:

```rust
pub enum MyEnum {
    First { a: u32, b: i32 },
    Second(u32),
}
```

This type's debug info is the following `DICompositeType` with `DW_TAG_structure_type` tag:

```llvm
!4 = !DICompositeType(tag: DW_TAG_structure_type, name: "MyEnum",
     scope: !2, file: !5, size: 96, align: 32, flags: DIFlagPublic,
     elements: !6, templateParams: !16,
     identifier: "faba668fd9f71e9b7cf3b9ac5e8b93cb")
```

With one element being also a `DICompositeType`, but with `DW_TAG_variant_part` tag:

```llvm
!6 = !{!7}
!7 = !DICompositeType(tag: DW_TAG_variant_part, scope: !4, file: !5,
     size: 96, align: 32, elements: !8, templateParams: !16,
     identifier: "e4aee046fc86d111657622fdcb8c42f7", discriminator: !21)
```

Which has a discriminator:

```llvm
!21 = !DIDerivedType(tag: DW_TAG_member, scope: !4, file: !5,
      baseType: !13, size: 32, align: 32, flags: DIFlagArtificial)
```

Which then holds different variants as `DIDerivedType` elements with `DW_TAG_member` tag:

```llvm
!8 = !{!9, !17}
!9 = !DIDerivedType(tag: DW_TAG_member, name: "First", scope: !7,
     file: !5, baseType: !10, size: 96, align: 32, extraData: i32 0)
!10 = !DICompositeType(tag: DW_TAG_structure_type, name: "First",
      scope: !4, file: !5, size: 96, align: 32, flags: DIFlagPublic,
      elements: !11, templateParams: !16,
      identifier: "cc7748c842e275452db4205b190c8ff7")
!11 = !{!12, !14}
!12 = !DIDerivedType(tag: DW_TAG_member, name: "a", scope: !10,
      file: !5, baseType: !13, size: 32, align: 32, offset: 32,
      flags: DIFlagPublic)
!13 = !DIBasicType(name: "u32", size: 32, encoding: DW_ATE_unsigned)
!14 = !DIDerivedType(tag: DW_TAG_member, name: "b", scope: !10,
      file: !5, baseType: !15, size: 32, align: 32, offset: 64,
      flags: DIFlagPublic)
!15 = !DIBasicType(name: "i32", size: 32, encoding: DW_ATE_signed)
!16 = !{}
!17 = !DIDerivedType(tag: DW_TAG_member, name: "Second", scope: !7,
      file: !5, baseType: !18, size: 96, align: 32, extraData: i32 1)
!18 = !DICompositeType(tag: DW_TAG_structure_type, name: "Second",
      scope: !4, file: !5, size: 96, align: 32, flags: DIFlagPublic,
      elements: !19, templateParams: !16,
      identifier: "a2094b1381f3082d504fbd0903aa7c06")
!19 = !{!20}
!20 = !DIDerivedType(tag: DW_TAG_member, name: "__0", scope: !18,
      file: !5, baseType: !13, size: 32, align: 32, offset: 32,
      flags: DIFlagPublic)
```

BPF backend was assuming that all the elements of any `DICompositeType` have tag `DW_TAG_member` and are instances of `DIDerivedType`. However, the single element of the outer composite type `!4` has tag `DW_TAG_variant_part` and is an instance of `DICompositeType`. The unconditional call of `cast<DIDerivedType>` on all elements was causing an assertion failure when any Rust code with enums was compiled to the BPF target.

Fix that by:

* Handling `DW_TAG_variant_part` in `visitStructType`.
* Replacing unconditional call of `cast<DIDerivedType>` over `DICompositeType` elements with a `switch` statement, handling both `DW_TAG_member` and `DW_TAG_variant_part` and casting the element to an appropriate type (`DIDerivedType` or `DICompositeType`).

Fixes: https://github.com/llvm/llvm-project/issues/155778

vadorovsky commented on llvm/test/CodeGen/BPF/BTF/data-carrying-enum.ll:
```diff
@@ -0,0 +1,110 @@
+;   }
+; Compilation flag:
+;   cargo +nightly rustc -Zbuild-std=core --target=bpfel-unknown-none -- --emit=llvm-bc
+;   llvm-extract --glob=X --glob=Y --glob=Z $(find target/ -name "*.bc" | head -n 1) -o data-carrying-enum.bc
```
This behavior can be reproduced only with Rust. However, this trick cleans up the noisy IR produces by Rust from the panic handler and all unnecessary core types. I hope it's good enough for the test.

vadorovsky committed:
[BPF] Support for `DW_TAG_variant_part` in BTF generation

Variant part, represented by `DW_TAG_variant_part` is a structure with a
discriminant and different variants, from which only one can be active
and valid at the same time. The discriminant is the main difference
between variant parts and unions represented by `DW_TAG_union` type.

Variant parts are used by Rust enums, which look like:

```rust
pub enum MyEnum {
    First { a: u32, b: i32 },
    Second(u32),
}
```

This type's debug info is the following `DICompositeType` with
`DW_TAG_structure_type` tag:

```llvm
!4 = !DICompositeType(tag: DW_TAG_structure_type, name: "MyEnum",
     scope: !2, file: !5, size: 96, align: 32, flags: DIFlagPublic,
     elements: !6, templateParams: !16,
     identifier: "faba668fd9f71e9b7cf3b9ac5e8b93cb")
```

With one element being also a `DICompositeType`, but with
`DW_TAG_variant_part` tag:

```llvm
!6 = !{!7}
!7 = !DICompositeType(tag: DW_TAG_variant_part, scope: !4, file: !5,
     size: 96, align: 32, elements: !8, templateParams: !16,
     identifier: "e4aee046fc86d111657622fdcb8c42f7", discriminator: !21)
```

Which has a discriminator:

```llvm
!21 = !DIDerivedType(tag: DW_TAG_member, scope: !4, file: !5,
      baseType: !13, size: 32, align: 32, flags: DIFlagArtificial)
```

Which then holds different variants as `DIDerivedType` elements with
`DW_TAG_member` tag:

```llvm
!8 = !{!9, !17}
!9 = !DIDerivedType(tag: DW_TAG_member, name: "First", scope: !7,
     file: !5, baseType: !10, size: 96, align: 32, extraData: i32 0)
!10 = !DICompositeType(tag: DW_TAG_structure_type, name: "First",
      scope: !4, file: !5, size: 96, align: 32, flags: DIFlagPublic,
      elements: !11, templateParams: !16,
      identifier: "cc7748c842e275452db4205b190c8ff7")
!11 = !{!12, !14}
!12 = !DIDerivedType(tag: DW_TAG_member, name: "a", scope: !10,
      file: !5, baseType: !13, size: 32, align: 32, offset: 32,
      flags: DIFlagPublic)
!13 = !DIBasicType(name: "u32", size: 32, encoding: DW_ATE_unsigned)
!14 = !DIDerivedType(tag: DW_TAG_member, name: "b", scope: !10,
      file: !5, baseType: !15, size: 32, align: 32, offset: 64,
      flags: DIFlagPublic)
!15 = !DIBasicType(name: "i32", size: 32, encoding: DW_ATE_signed)
!16 = !{}
!17 = !DIDerivedType(tag: DW_TAG_member, name: "Second", scope: !7,
      file: !5, baseType: !18, size: 96, align: 32, extraData: i32 1)
!18 = !DICompositeType(tag: DW_TAG_structure_type, name: "Second",
      scope: !4, file: !5, size: 96, align: 32, flags: DIFlagPublic,
      elements: !19, templateParams: !16,
      identifier: "a2094b1381f3082d504fbd0903aa7c06")
!19 = !{!20}
!20 = !DIDerivedType(tag: DW_TAG_member, name: "__0", scope: !18,
      file: !5, baseType: !13, size: 32, align: 32, offset: 32,
      flags: DIFlagPublic)
```

BPF backend was assuming that all the elements of any `DICompositeType`
have tag `DW_TAG_member` and are instances of `DIDerivedType`. However,
the single element of the outer composite type `!4` has tag
`DW_TAG_variant_part` and is an instance of `DICompositeType`. The
unconditional call of `cast<DIDerivedType>` on all elements was causing
an assertion failure when any Rust code with enums was compiled to the
BPF target.

Fix that by:

* Handling `DW_TAG_variant_part` in `visitStructType`.
* Replacing unconditional call of `cast<DIDerivedType>` over
  `DICompositeType` elements with a `switch` statement, handling
  both `DW_TAG_member` and `DW_TAG_variant_part` and casting the element
  to an appropriate type (`DIDerivedType` or `DICompositeType`).

To keep BTF simple and make BTF relocations correct, represent the
discriminator as the first element and apply an offset to all elements.

Fixes: #155778

eddyz87 commented on llvm/test/CodeGen/BPF/BTF/data-carrying-enum.ll:
```diff
@@ -0,0 +1,95 @@
+
+; CHECK-BTF:      [1] STRUCT 'DataCarryingEnum' size=12 vlen=1
+; CHECK-BTF-NEXT:         '(anon)' type_id=2 bits_offset=0
+; CHECK-BTF-NEXT: [2] STRUCT '(anon)' size=12 vlen=2
```
This should be a union, should it? Otherwise 'size' feild for [2] does not make sense: 12 vs 12 + 12.

vadorovsky replied:
Good point. After applying your other suggestions, it became an `UNION` without any further changes.

eddyz87 commented on llvm/test/CodeGen/BPF/BTF/data-carrying-enum.ll:
```diff
@@ -0,0 +1,95 @@
+; CHECK-BTF-NEXT:         'b' type_id=5 bits_offset=64
+; CHECK-BTF-NEXT: [4] INT 'u32' size=4 bits_offset=0 nr_bits=32 encoding=(none)
+; CHECK-BTF-NEXT: [5] INT 'i32' size=4 bits_offset=0 nr_bits=32 encoding=SIGNED
+; CHECK-BTF-NEXT: [6] STRUCT 'Second' size=12 vlen=2
```
Having a test case with 'Second' being of a different size would be more interesting, imo. The named vs unnamed members part matters only for rust frontend. At the backend level it does not matter.

eddyz87 commented:
Let's change the name to: `Support for DW_TAG_variant_part in BTF generation`. Data carrying enum is not an official term anyway (is this how algebraic data types are called these days?).

eddyz87 commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -305,17 +305,22 @@ void BTFTypeStruct::completeType(BTFDebug &BDebug) {
-      BTFMember.Offset = BitFieldSize << 24 | DDTy->getOffsetInBits();
-    } else {
-      BTFMember.Offset = DDTy->getOffsetInBits();
+    if (const auto *DDTy = dyn_cast<DIDerivedType>(Element)) {
```
Same here, I'd add explicit switch checking tags.

eddyz87 reviewed (changes requested):
@vadorovsky ,

Looks good except for the note about HasBitField from Yonghong. Do you plan to wrap this up?

vadorovsky commented:
> Do you plan to wrap this up?

I do, but I'm blocked by Rust which is emitting `undef` for all static variable definitions. Still trying to figure out a proper fix on Rust's side. If you prefer, I can close or mark this PR as draft until I sort it out.

vadorovsky commented:
Oh, so in your case rustc didn't emit `undef` in the constant, interesting. In my case, the IR fragments defining constants look like:

https://github.com/vadorovsky/llvm-project/commit/413a98a961841510f813244f454a34ef6e9f116a#diff-f411e8cbe94fed63916ca9eebac7652eddbb2db4770934eb94c26632216cbc58R42

The difference in our codes that I added more constants for different variants and added `no_mangle` attribute.

I'm AFK ATM. Tomorrow I will try to produce a similar IR without `undef`, or just use your IR in the test. What Rust version did you use? I used the latest nightly from a week ago.

(The issue with Rust emitting `undef` in any case will still will need to be fixed, eventually, but good to know it doesn't necessarily block my PR 🙂 )

eddyz87 commented:
> Oh, so in your case rustc didn't emit `undef` in the constant, interesting. In my case, the IR fragments defining constants look like:
> 
> [vadorovsky@413a98a#diff-f411e8cbe94fed63916ca9eebac7652eddbb2db4770934eb94c26632216cbc58R42](https://github.com/vadorovsky/llvm-project/commit/413a98a961841510f813244f454a34ef6e9f116a#diff-f411e8cbe94fed63916ca9eebac7652eddbb2db4770934eb94c26632216cbc58R42)
> 
> The difference in our codes that I added one more constant for the other variant and used `no_mangle` attribute.

It's a backend test. When I write such tests I start from something generated by frontend but cut away any unrelated stuff. In this particular case the only relevant thing is the structure of debug information generated for variant/variant_part. You can drop initialization from those globals and it would still be fine for the backend test.
 
> I'm AFK ATM. Tomorrow I will try to produce a similar IR without `undef`, or just use your IR in the test. What Rust version did you use? I used the latest nightly from a week ago.

rustc version 1.88.0 (6b00bc388 2025-06-23) (Red Hat 1.88.0-1.el9

eddyz87 reviewed (changes requested):
I think this change is fine, but just realized an unfortunate quirk regarding discriminator handling. Consider the following example:

```rust
pub enum Adt {
    First { a: u32, b: i32 },
    Second(u32, i32),
}

pub static X: Adt = Adt::First{a:0, b:0};
```

With corresponding IR:

```
!7 = !DICompositeType(tag: DW_TAG_variant_part, scope: !4, file: !5, size: 96, align: 32, elements: !8, ..., discriminator: !22)
  !8 = !{!9, !17}
  !9 = !DIDerivedType(tag: DW_TAG_member, name: "First", scope: !7, file: !5, baseType: !10, size: 96, align: 32, extraData: i32 0)
  !10 = !DICompositeType(tag: DW_TAG_structure_type, name: "First", ..., elements: !11, ...)
    !11 = !{!12, !14}
    !12 = !DIDerivedType(tag: DW_TAG_member, name: "a", scope: !10, file: !5, baseType: !13, size: 32, align: 32, offset: 32, flags: DIFlagPublic)
      ...

  !22 = !DIDerivedType(tag: DW_TAG_member, scope: !4, file: !5, baseType: !13, size: 32, align: 32, flags: DIFlagArtificial)
```

And corresponding DWARF:

```c
0x0000003d:     DW_TAG_structure_type
                  DW_AT_name    ("Adt")
                  DW_AT_byte_size       (0x0c)
                  DW_AT_accessibility   (DW_ACCESS_public)
                  DW_AT_alignment       (4)

0x00000045:       DW_TAG_variant_part
                    DW_AT_discr (0x0000004a)

0x0000004a:         DW_TAG_member
                      DW_AT_type        (0x000000b2 "u32")
                      DW_AT_alignment   (4)
                      DW_AT_data_member_location        (0x00)
                      DW_AT_artificial  (true)

0x00000051:         DW_TAG_variant
                      DW_AT_discr_value (0x00)

0x00000053:           DW_TAG_member
                        DW_AT_name      ("First")
                        DW_AT_type      (0x0000006e "test_debug_info::Adt::First")
                        DW_AT_alignment (4)
                        DW_AT_data_member_location      (0x00)

0x0000005e:           NULL

0x0000005f:         DW_TAG_variant
                      DW_AT_discr_value (0x01)

0x00000061:           DW_TAG_member
                        DW_AT_name      ("Second")
                        DW_AT_type      (0x0000008f "test_debug_info::Adt::Second")
                        DW_AT_alignment (4)
                        DW_AT_data_member_location      (0x00)

0x0000006c:           NULL

0x0000006d:         NULL

0x0000006e:       DW_TAG_structure_type
                    DW_AT_name  ("First")
                    DW_AT_byte_size     (0x0c)
                    DW_AT_accessibility (DW_ACCESS_public)
                    DW_AT_alignment     (4)

0x00000076:         DW_TAG_member
                      DW_AT_name        ("a")
                      DW_AT_type        (0x000000b2 "u32")
                      DW_AT_alignment   (4)
                      DW_AT_data_member_location        (0x04)
                      DW_AT_accessibility       (DW_ACCESS_public)

0x00000082:         DW_TAG_member
                      DW_AT_name        ("b")
                      DW_AT_type        (0x000000b9 "i32")
                      DW_AT_alignment   (4)
                      DW_AT_data_member_location        (0x08)
                      DW_AT_accessibility       (DW_ACCESS_public)

0x0000008e:         NULL
```

Note how offsets for `a` and `b` are shifted by 4 bytes and variant part itself has three members, beside First and Second there is also an anonymous member `0x0000004a`, representing a descriminator at offset 0.

Meaning that in BTF union representing the variant part has to have three members, not two. Which is a bit inconvenient, as it is not a part of "elements", but instead a separate "discriminator" reference.

Wdyt?

vadorovsky commented:
@eddyz87 @yonghong-song I got it working and went with the additional member solution.

vadorovsky commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -301,21 +303,62 @@ void BTFTypeStruct::completeType(BTFDebug &BDebug) {
+    struct BTF::BTFMember Discriminator;
+    const auto *DDTy = STy->getDiscriminator();
+
+    InitialOffset += DDTy->getOffsetInBits() + DDTy->getSizeInBits();
```
I think there is no chance for `getOffsetInBits()` being anything else than 0 here, but I'm still including it to make the implementation generic instead of just assuming things. But let me know if you have other thoughts.

yonghong-song commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -301,21 +303,62 @@ void BTFTypeStruct::completeType(BTFDebug &BDebug) {
+
+    Discriminator.NameOff = BDebug.addString(DDTy->getName());
+    Discriminator.Offset = DDTy->getOffsetInBits();
+    const auto *BaseTy = tryRemoveAtomicType(DDTy->getBaseType());
```
I do not think we need tryRemoveAtomicType(). Did you find a use case that tryRemoveAtomicType() could be necessary?

vadorovsky replied:
I didn't. You're right, I removed it.

yonghong-song commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -301,21 +303,62 @@ void BTFTypeStruct::completeType(BTFDebug &BDebug) {
+      BTFMember.NameOff = BDebug.addString(DCTy->getName());
+      BTFMember.Offset = InitialOffset + DCTy->getOffsetInBits();
+      const auto *DTy = cast<DIType>(DCTy);
+      BTFMember.Type = BDebug.getTypeId(DTy);
```
DTy is not needed. Directly using BDebug.getTypeId(DCTy).

yonghong-song commented on llvm/test/CodeGen/BPF/BTF/variant-part.ll:
```diff
@@ -0,0 +1,87 @@
+; CHECK-BTF-NEXT: [2] UNION '(anon)' size=12 vlen=3
+; CHECK-BTF-NEXT:         '(anon)' type_id=4 bits_offset=0
+; CHECK-BTF-NEXT:         'First' type_id=3 bits_offset=32
+; CHECK-BTF-NEXT:         'Second' type_id=6 bits_offset=32
```
I think we kindly agree that we do not like have different bit_offset for union members.

vadorovsky replied:
Would you agree then with representing it as a struct with a discriminator and an union? Like:

```
[2] STRUCT '(anon)' size=12 vlen=2
        '(anon)' type_id=4 bits_offset=0 // discriminator
        '(anon)' type_id=3 bits_offset=32 // union with variants
[3] UNION '(anon)' size=8 vlen=2
        'First' type_id=4 bits_offset=0
        'Second' type_id=7 bits_offset=0
```

That's what I proposed initially https://github.com/llvm/llvm-project/pull/155783#issuecomment-3286581654 and I'm quite confident that the code won't be too complicated.

vadorovsky commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -301,21 +303,61 @@ void BTFTypeStruct::completeType(BTFDebug &BDebug) {
+    struct BTF::BTFMember Discriminator;
+    const auto *DDTy = STy->getDiscriminator();
+
+    InitialOffset += DDTy->getOffsetInBits() + DDTy->getSizeInBits();
```
My worry is that we don't do the offset computation correctly, the BTF relocations done on variant parts won't work. Hypocritical of me given that I didn't get it right here (I'm very sorry, I'm still new to LLVM and compilers), but I really want to do it right. I'm planning to add BTF relocations for Rust BPF programs eventually (we want to emit them it in [bpf-linker](https://github.com/aya-rs/bpf-linker), given that we traverse patch bc/IR there) and it would be bad if I have to add yet another LLVM patch to fix it post-factum.

Regarding stability - I think the variant parts annotated with `#[repr(C)]` are guaranteed to be stable. `#[repr(Rust)]` could change the layout any time,  but I think it's fair to require the usage of `#[repr(C)]` in BPF programs.

eddyz87 replied:
> Regarding stability - I think the variant parts annotated with #[repr(C)] are guaranteed to be stable. #[repr(Rust)] could change the layout any time, but I think it's fair to require the usage of #[repr(C)] in BPF programs. We could sort of enforce that by adding a custom Clippy lint and push Aya users to adhere to it.

I'm not worried about binary stability, but about generated DWARF stability. I don't think rustc gives any guarantees regarding DWARF representation of enums. Specifically, for the example at hand DWARF describes an equivalent of C structure below:

```c
// repr A
union Adt {
  u32 discriminant;
  struct First First;
  struct Second Second;
};
struct First {
  u32 :32;
  u32 a;
  u32 b;
};
...
```

But an alternative representation would be correct as well: 

```c
// repr B
struct Adt {
  u32 discriminant;
  union {
    struct First First;
    struct Second Second;
  }
};
struct First {
  u32 a;
  u32 b;
};
...
```

What you are trying to do with `InitialOffset` is to convert repr A into repr B, which is recursive operation closely tied to the specifics of DWARF generation by rustc. Should rustc decide to use different DWARF representation the logic will need adjustment.

> My worry is that we don't do the offset computation correctly, the BTF relocations done on variant parts won't work. 

DWARF produced by rustc already contains correct offsets for data structure fields, what is it exactly that you suspect won't work?

> Hypocritical of me given that I didn't get it right here (I'm very sorry, I'm still new to LLVM and compilers), but I really want to do it right.

Nothing to be sorry about, compilers are complicated.

vadorovsky commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -301,21 +303,61 @@ void BTFTypeStruct::completeType(BTFDebug &BDebug) {
+    // discriminator as the first element and apply offsets to the other
+    // elements.
+    struct BTF::BTFMember Discriminator;
+    const auto *DDTy = STy->getDiscriminator();
```
Done

tamird replied:
I think this is not done?

yonghong-song commented on llvm/test/CodeGen/BPF/BTF/variant-part.ll:
```diff
@@ -0,0 +1,87 @@
+!llvm.dbg.cu = !{!27}
+
+; CHECK-BTF:      [1] STRUCT 'MyEnum' size=12 vlen=1
+; CHECK-BTF-NEXT:         '(anon)' type_id=3 bits_offset=0
```
Not sure whether this is an issue or not. Here, we have '(anon)' field name which essentially is a 'empty' name. If people goes to access 'First' struct from Struct 'MyEnum', the C access type will be like `MyEnum.(anon).First`,
could this be a problem or there is rust way to handle this?

Also without 'name' for discriminator field, not sure how it will be used by BTF/kernel etc.

eddyz87 commented:
For the reference: I checked BTF generation for kernel BPF selftests and there are (unsurprisingly) no differences with and w/o this patch. Should be ready to land once the remaining nits are addressed.

vadorovsky commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -672,16 +714,23 @@ void BTFDebug::visitStructType(const DICompositeType *CTy, bool IsStruct,
+  // element and instead keeps it as a separate reference. But we represent it
+  // as an element in BTF.
+  if (CTy->getTag() == dwarf::DW_TAG_variant_part)
+    VLen++;
```
Sorry, should be done for real. I also agree with @tamird's point about extracting the `getDiscriminator` call - done here and in one other place.
